src/middlewared/middlewared/plugins/acme_protocol.py

```diff
@@ -0,0 +1,366 @@
+from middlewared.validators import validate_attributes
+
+from acme import client, messages
+from botocore import exceptions as boto_exceptions
+from botocore.errorfactory import BaseClientExceptions as boto_BaseClientException
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+
+# TODO: See what can be done to respect rate limits
+
+
+class ACMERegistrationService(CRUDService):
+
+    class Config:
+        datastore = 'system.acmeregistration'
+        datastore_extend = 'acme.registration.register_extend'
+        namespace = 'acme.registration'
+        private = True
+
+    async def query(self, filters=None, options=None):
+        for filter in filters or []:
+            if len(filter) == 3 and filter[0] == 'directory':
+                # Let's normalize directory uri for query
+                filter[2] += '/' if filter[2][-1] != '/' else ''
+
+        return await super().query(filters, options)
+
+    async def register_extend(self, data):
+        data['body'] = {
+            key: value for key, value in
+            (await self.middleware.call(
+                'datastore.query', 'system.acmeregistrationbody',
+                [['acme', '=', data['id']]], {'get': True}
+            )).items() if key != 'acme'
+        }
+        return data
+
+    def get_directory(self, acme_directory_uri):
+        try:
+            response = requests.get(acme_directory_uri).json()
+            return messages.Directory({
+                key: response[key] for key in ['newAccount', 'newNonce', 'newOrder', 'revokeCert']
+            })
+        except (requests.ConnectionError, requests.Timeout, json.JSONDecodeError, KeyError) as e:
+            raise CallError(f'Unable to retrieve directory : {e}')
+
+    @accepts(
+        Dict(
+            'acme_registration_create',
+            Bool('tos', default=False),
+            Dict(
+                'JWK_create',
+                Int('key_size', default=2048),
+                Int('public_exponent', default=65537)
+            ),
+            Str('acme_directory_uri', required=True),
+        )
+    )
+    def do_create(self, data):
+
+        # STEPS FOR CREATION
+        # 1) CREATE KEY
+        # 2) REGISTER CLIENT
+        # 3) SAVE REGISTRATION OBJECT
+        # 4) SAVE REGISTRATION BODY
+
+        verrors = ValidationErrors()
+
+        directory = self.get_directory(data['acme_directory_uri'])
+        if not isinstance(directory, messages.Directory):
+            verrors.add(
+                'acme_registration_create.acme_directory_uri',
+                f'System was unable to retrieve the directory with the specified acme_directory_uri: {directory}'
+            )
+
+        # Normalizing uri after directory call as let's encrypt staging api
+        # does not accept a trailing slash right now
+        data['acme_directory_uri'] += '/' if data['acme_directory_uri'][-1] != '/' else ''
+
+        if not data['tos']:
```

Comment:
Are we exposing the ToS or a link to it in the UI? If not there is no point in providing it IMO.

We need to discuss implications.

Reply:
Yup, we are exposing it. It doesn't do much except for a simple check for the user that you need to be aware of terms of service of the acme server you are using